### PR TITLE
Consistent use of CordaException and CordaRuntimeException

### PR DESCRIPTION
**Pull request by mnesbit**
Custom exceptions in corda, should either derive from an appropriate closely related java exception, or CordaException, or CordaRuntimeException. They should not inherit just from Exception, or RuntimeException.

### CODE REVIEW

**Comment by mnesbit** [comment]
I raised this to make us more consistent. Personally I hate custom exceptions as they add very little, but we should at least be consistent if we have a CordaException class.

File: node/src/main/kotlin/net/corda/node/services/api/ServiceHubInternal.kt [hunk 1]
```diff
@@ -61,7 +62,7 @@ interface NetworkMapCacheInternal : NetworkMapCache {
 }
 
 @CordaSerializable
-sealed class NetworkCacheError : Exception() {
+sealed class NetworkCacheError : CordaException("Network Cache Error", null) {
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/api/ServiceHubInternal.kt]: These are not Errors.  How much turmoil is renaming to ...Exception?

File: node/src/main/kotlin/net/corda/node/services/network/NetworkMapService.kt [hunk 2]
```diff
@@ -360,7 +361,7 @@ class WireNodeRegistration(raw: SerializedBytes<NodeRegistration>, sig: DigitalS
 }
 
 @CordaSerializable
-sealed class NodeMapError : Exception() {
+sealed class NodeMapError : CordaException("Network Map Protocol Error", null) {
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/network/NetworkMapService.kt]: Same thing.  Shouldn't be Error.  If we rename to Exception, what's the damage?

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 3]
```diff
@@ -465,7 +466,7 @@ abstract class AbstractNode(open val configuration: NodeConfiguration,
     }
 
     // Specific class so that MockNode can catch it.
-    class DatabaseConfigurationException(msg: String) : Exception(msg)
+    class DatabaseConfigurationException(msg: String) : CordaException(msg, null)
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: Hadn't spotted this before.  Let's add another constructor to CordaException for the null cause case.  ie.  `constructor(message: String)`

File: client/rpc/src/main/kotlin/net/corda/client/rpc/PermissionException.kt [hunk 4]
```diff
@@ -1,10 +1,11 @@
...
+import net.corda.core.CordaRuntimeException
 import net.corda.core.serialization.CordaSerializable
 
 /**
  * Thrown to indicate that the calling user does not have permission for something they have requested (for example
  * calling a method).
  */
 @CordaSerializable
-class PermissionException(msg: String) : RuntimeException(msg)
+class PermissionException(msg: String) : CordaRuntimeException(msg)
```
Comment (vkolomeyko) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/PermissionException.kt]: Minor: `@CordaSerializable` annotation is no longer necessary as `CordaThrowable` already has it.

**Comment by vkolomeyko** [comment]
Is this change really necessary for 1.0 or can it wait till 1.1 ?

**Comment by mnesbit** [comment]
This is public API and these exceptions are for life. People are going to start catching them soon, especially things like notary exceptions, so that they can retry in race conditions.

File: client/rpc/src/main/kotlin/net/corda/client/rpc/PermissionException.kt (hunk 4 above)
Comment (mnesbit) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/PermissionException.kt]: I agree, but the window is closing, so I will leave and hope we marge as soon as this build is completed.

**Comment by mnesbit** [comment]
Missed the cut for 1.0

**Pull request merged**